http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1491,182 +1449,359 @@ private <I> Publisher<URI> resolveURI(io.micronaut.http.HttpRequest<I> request,
-            return Flux.error(decorate(new HttpClientResponseException(response.reason(), response)));
+            return Mono.error(decorate(new HttpClientResponseException(response.reason(), response)));
         } else {
-            return Flux.just(response);
+            return Mono.just(response);
         }
     }
 
-    private void streamRequestThroughChannel0(
+    /**
+     * This is the high-level request method. It sits above {@link #sendRawRequest} and handles
+     * things like filters, error handling, response parsing, request writing.
+     *
+     * @param parentRequest The parent <i>server</i> request from {@link ServerRequestContext}, for context propagation
+     * @param blockHint     The optional block hint
+     * @param request       The request to send. Must have resolved absolute URI (see {@link #resolveURI})
+     * @param readResponse  Function that reads the response from the raw
+     *                      {@link NettyClientByteBodyResponse} representation. This is run exactly
+     *                      once, but if there is a redirect, it potentially runs with a different
+     *                      request than the original (which is why it has a request parameter)
+     * @return A mono containing the response
+     */
+    private Mono<HttpResponse<?>> sendRequestWithRedirects(
         io.micronaut.http.HttpRequest<?> parentRequest,
+        @Nullable BlockHint blockHint,
         MutableHttpRequest<?> request,
-        FluxSink<? super MutableHttpResponse<?>> emitter,
-        ConnectionManager.PoolHandle poolHandle,
-        boolean sse) throws HttpPostRequestEncoder.ErrorDataEncoderException {
+        BiFunction<MutableHttpRequest<?>, NettyClientByteBodyResponse, ? extends Mono<? extends HttpResponse<?>>> readResponse
+    ) {
+        if (informationalServiceId != null && request.getAttribute(HttpAttributes.SERVICE_ID).isEmpty()) {
+            request.setAttribute(HttpAttributes.SERVICE_ID, informationalServiceId);
+        }
 
-        URI requestURI = request.getUri();
-        boolean permitsBody = io.micronaut.http.HttpMethod.permitsRequestBody(request.getMethod());
-        NettyByteBody byteBody = buildNettyRequest(
-            request,
-            requestURI,
-            request
-                .getContentType()
-                .orElse(MediaType.APPLICATION_JSON_TYPE),
-            permitsBody,
-            throwable -> {
-                if (!emitter.isCancelled()) {
-                    emitter.error(throwable);
-                }
-            },
-            poolHandle.channel.eventLoop()
-        );
-        String newUri = requestURI.getRawPath();
-        if (requestURI.getRawQuery() != null) {
-            newUri += "?" + requestURI.getRawQuery();
+        List<GenericHttpFilter> filters =
+            filterResolver.resolveFilters(request, clientFilterEntries);
+        if (parentRequest != null) {
+            // todo: migrate to new filter
```

Comment:
when will this todo be addressed?

Reply:
it's from the old code (applyFilterToResponsePublisher). it's a bit annoying to port to the new filter api since it's a programmatic filter and the new filter api is only declarative